src/assets/scripts/client/aircraft/FlightManagementSystem/LegModel.js

```diff
@@ -1,538 +1,780 @@
+        }
     }
 
     /**
-     * Find if a Waypoint exists within `#waypointCollection`
+     * Generate an array of `WaypointModel`s an aircraft's FMS will need in order to
+     * navigate along this leg instance.
      *
      * @for LegModel
-     * @method hasWaypoint
-     * @param waypointName {string}
-     * @return {boolean}
+     * @param _generateWaypointCollection
+     * @param entryOrFixName {string} name of the entry point (if airway/procedure), or fix name
+     * @param exit {string} name of exit point (if airway/procedure), or `undefined`
+     * @return {array<WaypointModel>}
      */
-    hasWaypoint(waypointName) {
-        // using a for loop here instead of `_find()` because this operation could happen a lot. a for
-        // loop is going to be faster than `_find()` in most cases.
-        for (let i = 0; i < this.waypointCollection.length; i++) {
-            const waypoint = this.waypointCollection[i];
+    _generateWaypointCollection(entryOrFixName, exit) {
+        if (this._legType === LEG_TYPE.DIRECT) {
+            return [new WaypointModel(entryOrFixName)];
+        }
 
-            if (waypointName.toLowerCase() === waypoint.name) {
-                return true;
-            }
+        if (this._legType === LEG_TYPE.AIRWAY) {
+            this._verifyAirwayAndEntryAndExitAreValid(entryOrFixName, exit);
+
+            return this._airwayModel.getWaypointModelsForEntryAndExit(entryOrFixName, exit);
         }
 
-        return false;
+        this._verifyProcedureAndEntryAndExitAreValid(entryOrFixName, exit);
+
+        return this._procedureModel.getWaypointModelsForEntryAndExit(entryOrFixName, exit);
     }
 
+    // ------------------------------ PUBLIC ------------------------------
+
     /**
-     * Collects the `#altitudeMaximum` value from each waypoint
-     * in the `#waypointCollection`, then finds the highest value
+     * Mark the specified waypoint as a hold waypoint
      *
      * @for LegModel
-     * @method getProcedureTopAltitude
-     * @return {number}
+     * @method activateHoldForWaypointName
+     * @param waypointName {string} name of waypoint in route
+     * @param holdParameters {object}
      */
-    getProcedureTopAltitude() {
-        const isMaximum = true;
+    activateHoldForWaypointName(waypointName, holdParameters) {
+        if (!this.hasWaypointName(waypointName)) {
+            return;
+        }
 
-        if (!this.isProcedure) {
-            return -1;
+        const waypointIndex = this._findIndexOfWaypointName(waypointName);
+        const waypointModel = this._waypointCollection[waypointIndex];
+
+        waypointModel.setHoldParametersAndActivateHold(holdParameters);
+    }
+
+    /**
+     * Return the ICAO identifier for the airport at which this leg will terminate (if
+     * it is in fact a STAR leg, of course).
+     *
+     * @for LegModel
+     * @method getArrivalRunwayAirportIcao
+     * @return {string}
+     */
+    getArrivalRunwayAirportIcao() {
```

Comment:
as written, we aren't _getting_ an airport icao here we are extracting it.

if this logic is extracted, like I mentioned previously, we should have easier access to this airport icao